File: Heaps/find_kth_largest.py
from heapq import heappop, heappush
from typing import List


class FindKthLargest:
    def findKthLargest(self, nums: List[int], k: int) -> int:
        min_heap = []
        for num in nums:
            heappush(min_heap, num)
            if len(min_heap) > k:
                heappop(min_heap)

        return min_heap[0]

# Time complexity: O(n log k) where n is the number of elements in nums and k is the value of k.
# We iterate through all n elements in nums, 
# and for each element, we perform a heap operation that takes O(log k) time. 
# After processing all elements, we return the root of the min heap, which is O(1).
# Space complexity: O(k) for the min heap that stores the k largest elements.

File: Heaps/FindKthLargest.java
package Heaps;

import java.util.PriorityQueue;

public class FindKthLargest {
    public int findKthLargest(int[] nums, int k) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        for (int num : nums) {
            minHeap.add(num);
            if (minHeap.size() > k)
                minHeap.remove();
        }

        return minHeap.peek();
    }
}

// Time Complexity: O(n log k) where n is the number of elements in the input
// array and k is the size of the heap. We add each element to the heap, which
// takes O(log k) time, and we do this for all n elements.
// Space Complexity: O(k) since we are maintaining a heap of size k.
